server/src/com/thoughtworks/go/server/materials/DependencyMaterialUpdateNotifier.java

```diff
@@ -0,0 +1,149 @@
+
+import java.util.*;
+
+import static java.lang.String.format;
+
+@Component
+public class DependencyMaterialUpdateNotifier implements StageStatusListener, ConfigChangedListener, Initializer, MaterialUpdateCompleteListener {
+    private static final Logger LOGGER = Logger.getLogger(DependencyMaterialUpdateNotifier.class);
+    private final GoConfigService goConfigService;
+    private final MaterialConfigConverter materialConfigConverter;
+    private final MaterialUpdateService materialUpdateService;
+
+    private Map<String, Material> dependencyMaterials;
+    private Set<Material> retryQueue = Collections.synchronizedSet(new HashSet<Material>());
+
+    @Autowired
+    public DependencyMaterialUpdateNotifier(GoConfigService goConfigService, MaterialConfigConverter materialConfigConverter,
+                                            MaterialUpdateService materialUpdateService) {
+        this.goConfigService = goConfigService;
+        this.materialConfigConverter = materialConfigConverter;
+        this.materialUpdateService = materialUpdateService;
+    }
+
+    public void initialize() {
+        loadDependencyMaterials();
+
+        goConfigService.register(this);
+        goConfigService.register(pipelineConfigChangedListener());
+        materialUpdateService.registerMaterialUpdateCompleteListener(this);
+
+        updateMaterialsOnIntialization();
+    }
+
+    @Override
+    public void onMaterialUpdate(Material material) {
+        if (material instanceof DependencyMaterial) {
+            if (retryQueue.remove(material)) {
+                if (LOGGER.isDebugEnabled()) {
+                    LOGGER.debug(format("[Material Update] Retrying update of dependency material %s ", material));
+                }
+                updateMaterial(material);
+            }
+        }
+    }
+
+    public void stageStatusChanged(Stage stage) {
+        if (StageResult.Passed == stage.getResult()) {
+            Material material = dependencyMaterials.get(stageIdentifier(stage.getIdentifier().getPipelineName(), stage.getName()));
+            if (material != null) {
+                updateMaterial(material);
+            }
+        }
+    }
+
+    @Override
+    public void onConfigChange(CruiseConfig newCruiseConfig) {
+        scheduleNewMaterialsForUpdate();
+    }
+
+    protected EntityConfigChangedListener<PipelineConfig> pipelineConfigChangedListener() {
+        final DependencyMaterialUpdateNotifier self = this;
+
+        return new EntityConfigChangedListener<PipelineConfig>() {
+            @Override
+            public void onEntityConfigChange(PipelineConfig pipelineConfig) {
+                self.scheduleNewMaterialsForUpdate();
+            }
+        };
+    }
+
+    private void updateMaterial(Material material) {
+        try {
+            if (!materialUpdateService.updateMaterial(material)) {
+                retryQueue.add(material);
+            }
+        } catch (Exception e) {
+            retryQueue.add(material);
```

Comment:
Added logging.

You are right, in case of an exception only the next run of pipeline will update this material. I thought of updating these as part of ```MaterialUpdateService.onTimer``` but that would take care of the retry scenario.